### ISSUE TITLE
[Apex] EmptyStatementBlock rule should provide options to ignore empty private constructors and empty virtual methods

### ISSUE THREAD

**Issue by SCWells72** [pmd/pmd#3568]
**Affects PMD Version:** 6.39.0

**Rule:** ApexDoc

Please provide the rule name and a link to the rule documentation:
https://pmd.github.io/pmd-6.39.0/pmd_rules_apex_errorprone.html#emptystatementblock

**Description:**

There are a few situations where empty blocks are actually expected and even support a best practices approach to the problem being solved, e.g.:

* `private` constructors in classes which should never be (externally) instantiated, e.g., singletons and utility classes
* default no-op implementations of `virtual` methods in `abstract` base classes where partial overrides are the norm

Options should be added to the `EmptyStatementBlock` rule to allow these specific types of instances to be ignored.

**Code Sample demonstrating the issue:**

**Singleton**
```apex
public with sharing class MySingleton {
    public static final MySingleton INSTANCE = new MySingleton();

    // This gets reported as an empty block, but it's critical to have a private constructor even when there's no constructor logic
    private MySingleton() {
        // Singleton
    }
}
```

**Utility class**
```apex
public with sharing class CollectionUtil {
    // This gets reported as an empty block, but it's critical to have a private constructor even when there's no constructor logic
    private CollectionUtil() {
        // Utility class
    }

    public static Boolean isEmpty(List<Object> values) { ... }
    public static Boolean isNotEmpty(List<Object> values) { ... }
    ...
}
```

**Abstract base class**
```apex
public abstract with sharing class AbstractTriggerHandler {
    // All of these are reported as empty blocks, but concrete subclasses will generally only override a subset of these methods,
    // so providing default no-op implementations is critical
    public virtual void beforeInsert() {}
    public virtual void afterInsert() {}
    public virtual void beforeUpdate() {}
    public virtual void afterUpdate() {}
    ...
}
```

**Expected outcome:**

I imagine for backward-compatibility, new options would be added for each of these categorical exceptions that default to the current behavior if unspecified, e.g.:

```xml
   <rule ref="category/apex/documentation.xml/EmptyStatementBlock" message="...">
      <priority>3</priority>
      <properties>
         <property name="reportEmptyPrivateConstructor" value="false" />
         <property name="reportEmptyVirtualMethod" value="false" />
      </properties>
   </rule>
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
CLI and direct API-based integration
